### ISSUE TITLE
#5139 looks broken (dropin loading)

### ISSUE THREAD

**Issue by poettering** [systemd/systemd#5248]
If I read #5139 correctly, it removes that we read the dropins of aliases. this doesn't look right...

@keszybz @fbuihuu?

**Comment by fbuihuu** [comment]
On 02/06/2017 09:24 PM, Lennart Poettering wrote:
> If I read #5139 <https://github.com/systemd/systemd/pull/5139>
> correctly, it removes that we read the dropins of aliases. this doesn't
> look right...
> 
> @keszybz @fbuihuu ?
> 

yes now with a better understanding, it seems broken.

But after restoring the old code, it will be still broken or
inconsistent IMHO.

As you can see the dependencies for a unit can vary:

 - if the aliases are already loaded for a given unit, then this
   unit when loaded in its turn will gain all dependencies defined
   through the aliases.

 - if none of the aliases are loaded then this specific unit will gain
   only the dependencies defined through its "true/final" unit name but
   not through its alias names.

It seems this inconsistency (and all the code that deal with unit
merging) is due to the way we load aliases: this is done lately and all
the aliases for a unit are not known after a call to
manager_load_unit_prepare().

I asked the question a couple of times now, so here it comes again: why
dont we resolve/search for all aliases for a given unit during
manager_load_unit_prepare() instead of "lazy" loading them ?

Thanks.

**Comment by poettering** [comment]
So, this is conceptually really fucked I think.

Think: b.service is an alias for a.service, but for the first 6h of system runtime, a.service is only referenced via a.service, and then a service is started that pulls in b.service, and now it is also referenced via b.service. You could argue that we'd now have to load all the b.service dropins too... But that's nothign we should do. I mean, loading unit settings should be a one time thing (except for deps).

I hence have the suspicion we should mandate that dropins via aliases are generally not permissible and we never check them.

**Comment by keszybz** [comment]
Sorry that I missed this change in behaviour. The code change looked correct...

> I hence have the suspicion we should mandate that dropins via aliases are generally not permissible and we never check them.

Yeah. I think that's a pragmatic solution: it's more likely for users to make the mistake of adding drop-ins for an alias thinking it's the main unit, then to want to have the configuration for a given unit split in two places.

**Comment by fsateler** [comment]
> Yeah. I think that's a pragmatic solution: it's more likely for users to make the mistake of adding drop-ins for an alias thinking it's the main unit, then to want to have the configuration for a given unit split in two places.

Also, this mistake is less likely now that `systemctl edit` follows link aliases (c6dd36b65caa9acc03a8e77b7db752574cf36eb4)

**Comment by fbuihuu** [comment]
> I hence have the suspicion we should mandate that dropins via aliases are generally not permissible and we never check them.

Well that would be odd from a user point of view.

I think all the troubles are coming from the way aliases are _lazy_ loaded.

So let me quote and ask this question again (probably the 4th time I'm asking ;) ):

> I asked the question a couple of times now, so here it comes again: why dont we resolve/search for all aliases for a given unit during manager_load_unit_prepare() instead of "lazy" loading them ?

**Comment by keszybz** [comment]
> why dont we resolve/search for all aliases for a given unit during manager_load_unit_prepare() instead of "lazy" loading them ?

That'd require searching all lookup paths for any files and reading all symlinks, every time we need to load a unit (or keeping some sort of cache). Not impossible, but not very elegant either.

**Comment by fbuihuu** [comment]
On 02/08/2017 02:06 PM, Zbigniew Jędrzejewski-Szmek wrote:
> That'd require searching all lookup paths for any files and reading all
> symlinks, every time we need to load a unit (or keeping some sort of
> cache). Not impossible, but not very elegant either.
> 

Well caching the symlinks in /*/systemd/system/ directories doesn't seem
that ugly at least it would make the handling of aliases much easier and
safer.

As you can see the current situation is not obvious and led to some
issues and inconsistencies.

And there's also all the code about unit merging which is hard to follow
and could be dropped.

My 2 cents.

**Comment by keszybz** [comment]
I guess that if we simply read all symlinks at reload time, it'd be pretty fast.
Yeah, the unit merging code is pretty hairy, but I don't think it could be eliminated completely with that: it is also applied to device units.

If you want to have a go at this, I'd be happy to review patches ;)

**Comment by poettering** [comment]
hmm, so far the whole design was to only read what we need, but not iterate through all dirs completely... that'd be quite a substantial change. not sure i like it too much

**Comment by keszybz** [comment]
I *feel* the same way, but rationally, we have a small number of aliases (33 on my Fedora 25 system), and ~400 unit files. Listing them and reading aliases up-front wouldn't be noticeable. If doing this would simplify handling of units, it certainly needs thinking about.

**Comment by fbuihuu** [comment]
On 02/09/2017 01:58 PM, Zbigniew Jędrzejewski-Szmek wrote:
> I /feel/ the same way, but rationally, we have a small number of aliases
> (33 on my Fedora 25 system), and ~400 unit files. Listing them and
> reading aliases up-front wouldn't be noticeable. If doing this would
> simplify handling of units, it certainly needs thinking about.

It wouldn't only simplify handling of units but also would remove the
limitations we're amending regarding the handling of dropin defined
through alias names.

**Comment by poettering** [comment]
so, i'll drop this issue from v233, this is not clear to me that any easy fix would be any better than the current master. we can investigate this later on again.

**Comment by keszybz** [comment]
Fixed by e8630e695232bdfcd16b55f3faafb4329c961104. If there's anything still wrong, let's open a new ticket.
